### PR TITLE
feat(hotspot): add a function to start hotkey detecting in shell commands 

### PR DESCRIPTION
**Pull request by Smityz**
### What problem does this PR solve? 
add a function to start and stop hotkey detecting in shell commands 

Because of the file structure, it's difficult to reuse code in `hotspot_partition_calculator::send_hotkey_detect_request`, so I reimplemented a code in client to send the RPC

### CODE REVIEW

File: src/shell/commands/data_operations.cpp [hunk 1]
```diff
@@ -2664,3 +2665,45 @@ bool calculate_hash_value(command_executor *e, shell_context *sc, arguments args
...
+// TODO: (Tangyanzhao) merge hotspot_partition_calculator::send_hotkey_detect_request
+bool detect_hotkey(command_executor *e, shell_context *sc, arguments args)
+{
+    static struct option long_options[] = {{"partition_num", required_argument, 0, 'p'},
+                                           {"hotkey_type", required_argument, 0, 't'},
+                                           {"action", required_argument, 0, 'a'},
+                                           {0, 0, 0, 0}};
+
+    std::string hotkey_type, hotkey_action;
+    int32_t partition_num = 0;
```
Comment (levy5307) [inline src/shell/commands/data_operations.cpp:2678]: uint32_t

File: src/client_lib/pegasus_client_impl.cpp [hunk 2]
```diff
@@ -1302,5 +1303,61 @@ const char *pegasus_client_impl::get_error_string(int error_code) const
...
+{
+    if (partition_hash < 0) {
+        err_info = "error partition_num";
+        return false;
+    }
+
+    dsn::apps::hotkey_detect_request req;
+    std::string hotkey_type_check = hotkey_type;
+    std::transform(
+        hotkey_type_check.begin(), hotkey_type_check.end(), hotkey_type_check.begin(), ::tolower);
```
Comment (levy5307) [inline src/client_lib/pegasus_client_impl.cpp:1320]: remove the prefix `::`

File: src/client_lib/pegasus_client_impl.cpp [hunk 3]
```diff
@@ -1302,5 +1303,61 @@ const char *pegasus_client_impl::get_error_string(int error_code) const
...
+                   ::tolower);
+    if (hotkey_action_check == "start") {
+        req.action = dsn::apps::hotkey_detect_action::START;
+    } else if (hotkey_action_check == "stop") {
+        req.action = dsn::apps::hotkey_detect_action::STOP;
+    } else {
+        err_info = "wrong action type";
+        return false;
+    }
+
```
Comment (Shuo-Jia) [inline src/client_lib/pegasus_client_impl.cpp:1343]: you‘d better refactor param `check` and `transform` to one method,  such as `generate_hotkey_request`. And I think it should be the `command` logic but not put into pegasus_client_iml.cpp

File: src/shell/commands/data_operations.cpp [hunk 4]
```diff
@@ -2664,3 +2665,45 @@ bool calculate_hash_value(command_executor *e, shell_context *sc, arguments args
...
+            break;
+        switch (c) {
+        case 'p':
+            partition_num = boost::lexical_cast<int32_t>(optarg);
+            break;
+        case 't':
+            hotkey_type = optarg;
+            break;
+        case 'a':
+            hotkey_action = optarg;
```
Comment (Shuo-Jia) [inline src/shell/commands/data_operations.cpp:2694]: Base previous comment, you should check and transfer param here, and that `detect_hotkey` only "detect_hotkey"

Comment (Smityz) [reply src/shell/commands/data_operations.cpp:2694]: The file structure here makes me a little confused, previous [codes](https://github.com/apache/incubator-pegasus/blob/master/src/client_lib/pegasus_client_impl.cpp#L670), there are also some params checkers

File: src/client_lib/pegasus_client_impl.h [hunk 5]
```diff
@@ -221,6 +221,10 @@ class pegasus_client_impl : public pegasus_client
                                  const scan_options &options,
                                  async_get_unordered_scanners_callback_t &&callback) override;
 
+    virtual bool detect_hotkey(const std::string &hotkey_type,
```
Comment (acelyc111) [inline src/client_lib/pegasus_client_impl.h:224]: Better to use dsn::apps::hotkey_type::type and dsn::apps::hotkey_detect_action as parameter, then no need to parse from string.

File: src/include/rrdb/rrdb.code.definition.h [hunk 6]
```diff
@@ -18,5 +18,6 @@ DEFINE_STORAGE_READ_RPC_CODE(RPC_RRDB_RRDB_TTL)
 DEFINE_STORAGE_READ_RPC_CODE(RPC_RRDB_RRDB_GET_SCANNER)
 DEFINE_STORAGE_READ_RPC_CODE(RPC_RRDB_RRDB_SCAN)
 DEFINE_STORAGE_READ_RPC_CODE(RPC_RRDB_RRDB_CLEAR_SCANNER)
+DEFINE_STORAGE_READ_RPC_CODE(RPC_DETECT_HOTKEY)
```
Comment (hycdong) [inline src/include/rrdb/rrdb.code.definition.h:21]: `DEFINE_STORAGE_READ_RPC_CODE` will define a rpc code whose action is read data from storage, start or stop detecting hotkey is not this rpc, it should not be defined here.

File: src/include/rrdb/rrdb.client.h [hunk 7]
```diff
@@ -417,6 +417,17 @@ class rrdb_client
...
 
+    // ---------- call RPC_DETECT_HOTKEY ------------
+    // - synchronous
+    std::pair<::dsn::error_code, hotkey_detect_response>
+    detect_hotkey_sync(const hotkey_detect_request &args,
+                       std::chrono::milliseconds timeout,
+                       uint64_t partition_hash)
+    {
+        return ::dsn::rpc::wait_and_unwrap<hotkey_detect_response>(_resolver->call_op(
+            RPC_DETECT_HOTKEY, args, &_tracker, empty_rpc_handler, timeout, partition_hash));
```
Comment (hycdong) [inline src/include/rrdb/rrdb.client.h:428]: RPC_DETECT_HOTKEY should not be a rrdb task code, see comments in `rrdb.code.definition.h`

File: src/shell/commands/data_operations.cpp (hunk 4 above)
Comment (Shuo-Jia) [reply src/shell/commands/data_operations.cpp:2694]: May just as the @acelyc111 comment: https://github.com/apache/incubator-pegasus/pull/605#discussion_r492069310, you'd better parse the string to `dsn::apps::hotkey_type::type` and `dsn::apps::hotkey_detect_action` here, and pass `detect_hotkey`

File: src/shell/commands/detect_hotkey.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                   "Hotkey detect rpc sending failed, in {}.{}, error_hint:{} {} \n",
+                   app_name,
+                   partition_index,
+                   rpc.response().err,
+                   rpc.response().err_hint);
+        return false;
+    }
+    std::cout << hotkey_action << " detecting hotkey succeed" << std::endl;
+    return true;
+}
```
Comment (levy5307) [inline src/shell/commands/detect_hotkey.cpp:116]: add a blank line

File: src/shell/commands/detect_hotkey.cpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                                app_name,
+                                partition_index,
+                                error.to_string());
+                 }
+             })
+        ->wait();
+    if (!is_successful) {
+        return false;
+    }
+    if (rpc.response().err != dsn::ERR_OK) {
```
Comment (levy5307) [inline src/shell/commands/detect_hotkey.cpp:105]: Is this `if` statement is repeated with L92~L98?

File: src/shell/commands/detect_hotkey.cpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    for (const auto &param : cmd.params()) {
+        if (params.find(param.first) == params.end()) {
+            fmt::print(stderr, "unknown param {} = {}\n", param.first, param.second);
+            return false;
+        }
+    }
+
+    return true;
+}
```
Comment (Shuo-Jia) [inline src/shell/commands/detect_hotkey.cpp:20]: The code is repeated with `disk_rebalance.cpp`, you can move it to `command_helper.h`

File: src/shell/commands/detect_hotkey.cpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
+#include "shell/commands.h"
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:1]: License

File: src/shell/main.cpp [hunk 12]
```diff
@@ -474,6 +474,15 @@ static command_executor commands[] = {
         "<-a --app_name str> [-f --forced]",
         cancel_bulk_load,
     },
+    {
+        "detect_hotkey",
+        "start or stop hotkey detection on the replica",
+        "[-a|--app_name str]"
+        "[-p|--partition_index num]"
+        "[-t|--hotkey_type read|write]"
+        "[-c|--detect_action start|stop]",
```
Comment (hycdong) [inline src/shell/main.cpp:483]: `<>` to show required parameters, and `[]` is for optional parameters. I suggest using `<>`.

File: src/shell/commands/detect_hotkey.cpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    std::transform(hotkey_action_check.begin(),
+                   hotkey_action_check.end(),
+                   hotkey_action_check.begin(),
+                   ::tolower);
+    if (hotkey_type_check == "read") {
+        req.type = dsn::apps::hotkey_type::type::READ;
+    } else if (hotkey_type_check == "write") {
+        req.type = dsn::apps::hotkey_type::type::WRITE;
+    } else {
+        fmt::print(stderr, "\"{}\"  is a invalid hotkey type (read or write) \n", hotkey_type);
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:39]:
```suggestion
        fmt::print(stderr, "\"{}\" is an invalid hotkey type (should be 'read' or 'write')\n", hotkey_type);
```

File: src/shell/commands/detect_hotkey.cpp [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        return false;
+    }
+    if (hotkey_action_check == "start") {
+        req.action = dsn::apps::hotkey_detect_action::START;
+    } else if (hotkey_action_check == "stop") {
+        req.action = dsn::apps::hotkey_detect_action::STOP;
+    } else {
+        fmt::print(
+            stderr, "\"{}\"  is a invalid hotkey detect action (start or stop) \n", hotkey_action);
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:48]:
```suggestion
            stderr, "\"{}\" is an invalid hotkey detect action (should be 'start' or 'stop')\n", hotkey_action);
```

File: src/shell/commands/detect_hotkey.cpp [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        "a", "app_name", "p", "partition_index", "c", "hotkey_action", "t", "hotkey_type"};
+    argh::parser cmd(args.argc, args.argv, argh::parser::PREFER_PARAM_FOR_UNREG_OPTION);
+    if (!validate_cmd(cmd, params)) {
+        return false;
+    }
+
+    std::string app_name = cmd({"-a", "--app_name"}).str();
+    uint64_t partition_index;
+    try {
+        partition_index = boost::lexical_cast<uint64_t>(cmd({"-p", "--partition_index"}).str());
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:68]: Use `dsn::buf2int64`

File: src/shell/commands/detect_hotkey.cpp [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+                 if (error != dsn::ERR_OK) {
+                     is_successful = false;
+                     fmt::print(stderr,
+                                "Hotkey detect rpc sending failed, in {}.{}, error_hint:{} \n",
+                                app_name,
+                                partition_index,
+                                error.to_string());
+                 }
+             })
+        ->wait();
```
Comment (hycdong) [inline src/shell/commands/detect_hotkey.cpp:101]: If partition_index is an invalid parameter, will it get_resolver succeed and send rpc succeed?
For example, this app partiion_count is 16, but partition_index is 18.

File: src/server/pegasus_read_service.h [hunk 17]
```diff
@@ -79,6 +81,7 @@ class pegasus_read_service : public dsn::replication::replication_app_base,
         register_rpc_handler_with_rpc_holder(dsn::apps::RPC_RRDB_RRDB_SCAN, "scan", on_scan);
         register_async_rpc_handler(
             dsn::apps::RPC_RRDB_RRDB_CLEAR_SCANNER, "clear_scanner", on_clear_scanner);
+        register_rpc_handler_with_rpc_holder(RPC_DETECT_HOTKEY, "detect_hotkey", on_detect_hotkey);
```
Comment (hycdong) [inline src/server/pegasus_read_service.h:84]: I still don't recommend that the `detect_hotkey` rpc handler is registered in `pegasus_read_service`.

File: src/shell/commands/detect_hotkey.cpp [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    dsn::task_tracker tracker;
+    pegasus::server::detect_hotkey_rpc rpc(
+        std::move(request_ptr), RPC_DETECT_HOTKEY, std::chrono::seconds(10), partition_index);
+    rpc.call(resolver,
+             &tracker,
+             [app_name, partition_index, &is_successful](dsn::error_code error) {
+                 if (error != dsn::ERR_OK) {
+                     is_successful = false;
+                     fmt::print(stderr,
+                                "Hotkey detect rpc sending failed, in {}.{}, error_hint:{} \n",
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:95]:
```suggestion
                                "Hotkey detect rpc sending failed, in {}.{}, error_hint:{}\n",
```

File: src/shell/commands/detect_hotkey.cpp [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    auto request_ptr = std::make_unique<dsn::apps::hotkey_detect_request>(req);
+    auto resolver = partition_resolver::get_resolver(
+        sc->current_cluster_name.c_str(), sc->meta_list, app_name.c_str());
+    bool is_successful = true;
+    dsn::task_tracker tracker;
+    pegasus::server::detect_hotkey_rpc rpc(
+        std::move(request_ptr), RPC_DETECT_HOTKEY, std::chrono::seconds(10), partition_index);
+    rpc.call(resolver,
+             &tracker,
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:90]: Is it needed to track this task even if you `wait` on it?

File: src/server/pegasus_server_impl.cpp [hunk 20]
```diff
@@ -2798,5 +2798,7 @@ void pegasus_server_impl::set_ingestion_status(dsn::replication::ingestion_statu
     _ingestion_status = status;
 }
 
+void pegasus_server_impl::on_detect_hotkey(detect_hotkey_rpc rpc) {}
```
Comment (acelyc111) [inline src/server/pegasus_server_impl.cpp:2801]: Add TODO then reviewers know you will implement it later.

File: src/shell/validate_utils.h [hunk 21]
```diff
@@ -0,0 +1,29 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (Shuo-Jia) [inline src/shell/validate_utils.h:1]: you can consider remove the header and rename the `validate_utils.cpp`  to `validate_utils.h directly`

File: src/shell/command_helper.h [hunk 22]
```diff
@@ -35,6 +35,7 @@
 
 #include "command_executor.h"
 #include "command_utils.h"
+#include "validate_utils.h"
```
Comment (levy5307) [inline src/shell/command_helper.h:38]: why did you add `#include "validate_utils.h"` here?

File: src/shell/validate_utils.cpp [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+    if (error != dsn::ERR_OK) {
+        err_info = fmt::format("list nodes failed, error={} \n", error.to_string());
+        return false;
+    }
+
+    bool not_find_ip = true;
+    for (const auto &node : nodes) {
+        if (ip_str == node.first.to_std_string()) {
+            target_address = node.first;
+            not_find_ip = false;
```
Comment (levy5307) [inline src/shell/validate_utils.cpp:67]: add break

File: src/shell/commands/detect_hotkey.cpp [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+                                           "t",
+                                           "hotkey_type",
+                                           "d",
+                                           "address"};
+    const std::set<std::string> flags = {};
+    argh::parser cmd(args.argc, args.argv, argh::parser::PREFER_PARAM_FOR_UNREG_OPTION);
+    if (!validate_cmd(cmd, params, flags)) {
+        return false;
+    }
+    int app_id;
```
Comment (levy5307) [inline src/shell/commands/detect_hotkey.cpp:79]: add a blank line

File: src/shell/commands/detect_hotkey.cpp [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+    }
+    int partition_index;
+    if (!dsn::buf2int32(cmd({"-p", "--partition_index"}).str(), partition_index)) {
+        fmt::print(stderr,
+                   "\"{}\" is a invalid partition index, you should use `app` to check related "
+                   "information\n",
+                   cmd({"-p", "--partition_index"}).str());
+        return false;
+    }
+    std::string hotkey_action = cmd({"-c", "--hotkey_action"}).str();
```
Comment (levy5307) [inline src/shell/commands/detect_hotkey.cpp:95]: move L95-L96 in front of L105

File: src/shell/validate_utils.h [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "shell/argh.h"
+#include <dsn/dist/fmt_logging.h>
```
Comment (levy5307) [inline src/shell/validate_utils.h:19]: use predeclare, so you can remove `#include command_executor.h`

Comment (levy5307) [inline src/shell/validate_utils.h:19]: remove `#include <dsn/dist/fmt_logging.h>`

File: src/shell/validate_utils.h (hunk 21 above)
Comment (levy5307) [reply src/shell/validate_utils.h:1]: I don't think it is a good idea. use `.h` and predeclare and avoid files which include `validate_utils.h` to dependency with `command_executor.h`

File: src/shell/validate_utils.h [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+#include "command_executor.h"
+
+bool validate_cmd(const argh::parser &cmd,
+                  const std::set<std::string> &params,
+                  const std::set<std::string> &flags);
```
Comment (hycdong) [inline src/shell/validate_utils.h:24]: `validate_cmd` can be implemented in `command_utis.h`, besides, add comments for this function.

File: src/shell/commands/detect_hotkey.cpp [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+        return false;
+    }
+    req.pid = dsn::gpid(app_id, partition_index);
+    return true;
+}
+
+// TODO: (Tangyanzhao) merge hotspot_partition_calculator::send_detect_hotkey_request
+bool detect_hotkey(command_executor *e, shell_context *sc, arguments args)
+{
+    // detect_hotkey [-a|--app_id] [-p|--partition_index][-c|--hotkey_action][-t|--hotkey_type]
```
Comment (hycdong) [inline src/shell/commands/detect_hotkey.cpp:63]: detect_hotkey [-a|--app_id] [-p|--partition_index][-c|--hotkey_action][-t|--hotkey_type][-d|--address]

File: src/shell/commands/detect_hotkey.cpp [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+{
+    if (!strcasecmp(hotkey_type.c_str(), "read")) {
+        req.type = dsn::replication::hotkey_type::type::READ;
+    } else if (!strcasecmp(hotkey_type.c_str(), "write")) {
+        req.type = dsn::replication::hotkey_type::type::WRITE;
+    } else {
+        err_info = fmt::format("\"{}\" is an invalid hotkey type (should be 'read' or 'write')\n",
+                               hotkey_type);
+        return false;
+    }
```
Comment (levy5307) [inline src/shell/commands/detect_hotkey.cpp:38]: add a blank line

File: src/shell/commands/disk_rebalance.cpp [hunk 30]
```diff
@@ -5,45 +5,14 @@
 #include "shell/commands.h"
 #include "shell/argh.h"
 #include "shell/command_output.h"
+#include "shell/command_helper.h"
```
Comment (levy5307) [inline src/shell/commands/disk_rebalance.cpp:8]: could you remove `#include "shell/command_helper.h"`?

Comment (Shuo-Jia) [reply src/shell/commands/disk_rebalance.cpp:8]: I think it seem should include `command_utils.h` which contains `validate_cmd` method

File: src/shell/commands/detect_hotkey.cpp [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+            stderr,
+            "\"{}\" is a invalid app_id, you should use `app` to check related information\n",
+            cmd({"-a", "--app_id"}).str());
+        return false;
+    }
+
+    int partition_index;
+    if (!dsn::buf2int32(cmd({"-p", "--partition_index"}).str(), partition_index)) {
+        fmt::print(stderr,
+                   "\"{}\" is a invalid partition index, you should use `app` to check related "
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:86]:
```suggestion
                   "\"{}\" is an invalid partition index, you should use `app` to check related "
```

File: src/shell/commands/detect_hotkey.cpp [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+    argh::parser cmd(args.argc, args.argv, argh::parser::PREFER_PARAM_FOR_UNREG_OPTION);
+    if (!validate_cmd(cmd, params, flags)) {
+        return false;
+    }
+
+    int app_id;
+    if (!dsn::buf2int32(cmd({"-a", "--app_id"}).str(), app_id)) {
+        fmt::print(
+            stderr,
+            "\"{}\" is a invalid app_id, you should use `app` to check related information\n",
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:78]: Update these logs, you only checked whether it's int type.

File: src/shell/commands/detect_hotkey.cpp [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+                   "\"{}\" is a invalid partition index, you should use `app` to check related "
+                   "information\n",
+                   cmd({"-p", "--partition_index"}).str());
+        return false;
+    }
+
+    dsn::rpc_address target_address;
+    std::string ip_str = cmd({"-d", "--address"}).str();
+    if (!target_address.from_string_ipv4(ip_str.c_str())) {
+        fmt::print("invalid ip, error={} \n", ip_str);
```
Comment (acelyc111) [inline src/shell/commands/detect_hotkey.cpp:95]:
```suggestion
        fmt::print("invalid ip, error={}\n", ip_str);
```

File: src/shell/main.cpp [hunk 34]
```diff
@@ -474,6 +474,16 @@ static command_executor commands[] = {
         "<-a --app_name str> [-f --forced]",
         cancel_bulk_load,
     },
+    {
+        "detect_hotkey",
+        "start or stop hotkey detection on the replica",
+        "<-a|--app_id str> "
```
Comment (acelyc111) [inline src/shell/main.cpp:480]:
```suggestion
        "<-a|--app_id num> "
```

File: src/shell/main.cpp [hunk 35]
```diff
@@ -474,6 +474,16 @@ static command_executor commands[] = {
         "<-a --app_name str> [-f --forced]",
         cancel_bulk_load,
     },
+    {
+        "detect_hotkey",
+        "start or stop hotkey detection on the replica",
```
Comment (acelyc111) [inline src/shell/main.cpp:479]:
```suggestion
        "start or stop hotkey detection on a replica of a replica server",
```

**Review by acelyc111 — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Review by hycdong — APPROVED** [review]
